common/context/src/main/java/io/helidon/common/context/ListContext.java

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright (c) 2019 Oracle and/or its affiliates. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.helidon.common.context;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Supplier;
+
+/**
+ * A {@link Context} implementation with deque registry.
+ */
+class ListContext implements Context {
+
+    private final Context parent;
+    private final ConcurrentHashMap<Object, ClassifiedRegistry> classifiers = new ConcurrentHashMap<>();
+    private final ClassifiedRegistry registry = new ClassifiedRegistry();
+
+    /**
+     * Creates new instance with defined parent.
+     *
+     * @param parent a parent context or {@code null}.
+     */
+    ListContext(Context parent) {
+        this.parent = parent;
+    }
+
+    /**
+     * Creates new instance.
+     */
+    ListContext() {
+        this(null);
+    }
+
+    @Override
+    public <T> void register(T instance) {
+        registry.register(instance);
+    }
+
+    @Override
+    public <T> void supply(Class<T> type, Supplier<T> supplier) {
+        registry.supply(type, supplier);
+    }
+
+    @Override
+    public <T> Optional<T> get(Class<T> type) {
+        T result = registry.get(type);
+        if (result == null) {
+            if (parent == null) {
+                return Optional.empty();
+            } else {
+                return parent.get(type);
+            }
+        } else {
+            return Optional.of(result);
+        }
+    }
+
+    @Override
+    public <T> void register(Object classifier, T instance) {
+        Objects.requireNonNull(classifier, "Parameter 'classifier' is null!");
```

Comment:
My preference is to statically import methods like this to declutter the code and make it more readable; don't know if we have an established practice here or not.

Reply:
Added to #648 , I disagree with you. Static imports may be fine, but sometimes they hide the source of the method and make it harder to read the code.